activedata_etl/transforms/per_test_to_es.py

```diff
@@ -0,0 +1,220 @@
+        record.test = {
+            "name": test,
+        }
+
+        # orphan lines (i.e. lines without a method), initialized to all lines
+        orphan_covered = set(covered)
+        orphan_uncovered = set(uncovered)
+
+        # iterate through the methods of this source file
+        # a variable to count the number of lines so far for this source file
+        methods = sf['functions'] if 'functions' in sf else []
+        method_start_indexes = [method['start'] for method in methods]
+        end = len(sf["coverage"])
+        for method in methods:
+            func_start = method['start']
+            func_end = end
+
+            for start in method_start_indexes:
+                if start > func_start:
+                    func_end = start
+                    break
+
+            method_lines = []
+            for l in coverable:
+                if l < func_start:
+                    continue
+
+                if l >= func_end:
+                    break
+
+                method_lines.append(l)
+
+            all_method_lines = set(method_lines)
+            method_covered = all_method_lines & file_covered
+            method_uncovered = all_method_lines - method_covered
+            method_percentage_covered = len(method_covered) / len(all_method_lines)
+
+            orphan_covered = orphan_covered - method_covered
+            orphan_uncovered = orphan_uncovered - method_uncovered
+
+            # Record method coverage info
+            record.source.method = {
+                "name": method['name'],
+                "covered": sorted(method_covered),
+                "uncovered": sorted(method_uncovered),
+                "total_covered": len(method_covered),
+                "total_uncovered": len(method_uncovered),
+                "percentage_covered": method_percentage_covered,
+            }
+
+            # Timestamp this record
+            record.etl.timestamp = Date.now()
+
+            key = etl2key(record.etl)
+            yield {"id": key, "value": record}
+
+        # a record for all the lines that are not in any method
+        # every file gets one because we can use it as canonical representative
+        # Record method coverage info
+        record.source.method = {
+            "covered": sorted(orphan_covered),
+            "uncovered": sorted(orphan_uncovered),
+            "total_covered": len(orphan_covered),
+            "total_uncovered": len(orphan_uncovered),
+            "percentage_covered": len(orphan_covered) / max(1, (len(orphan_covered) + len(orphan_uncovered))),
+        }
+
+        # Timestamp this record
+        record.etl.timestamp = Date.now()
+
+        key = etl2key(record.etl)
+        yield {"id": key, "value": record}
+
+    def generator():
+        with ZipFile(per_test_file) as zipped:
+            for zip_name in zipped.namelist():
+                with zipped.open(zip_name) as f:
+                    data = json.load(f)
```

Comment:
how big do these get?  I maybe we can stream the `source_files`?

Reply:
The `mo_json.stream.parse()` will accept a stream, a path to the (multidimensional) array being iterated, and a set of properties to return for each object in the array.  It is a bit slower, but it does reduce total memory:

     stream.parse
          80meg, 65sec
          71meg, 62sec
          84meg, 64sec
     
     json.load
          160meg, 57sec
          165meg, 55sec
          169meg, 54sec



https://github.com/klahnakoski/mo-json#method-mo_jsonstreamparse


```python
    def generator():
        with ZipFile(per_test_file) as zipped:
            for zip_name in zipped.namelist():
                for record in stream.parse(zipped.open(zip_name), "report.source_files", {"report.source_files", "suite", "test"}):
                    if please_stop:
                        Log.error("Shutdown detected. Stopping job ETL.")

                    try:
                        for d in process_source_file(
                            artifact_etl,
                            counter,
                            record.test,
                            record.report.source_files
                        ):
                            yield d
                    except Exception as e:
                        Log.warning(
                            "Error processing test {{test}} and source file {{source}} while processing {{key}}",
                            key=source_key,
                            test=record.test,
                            source=record.report.source_files.name,
                            cause=e
                        )
```

Reply:
Done!

On my machine parsing the test report with stream takes ~50 seconds, parsing with json.load takes ~35 seconds.